cpp/src/arrow/python/inference.cc

```diff
@@ -161,7 +169,7 @@ class NumPyDtypeUnifier {
       _NUMPY_UNIFY_NOOP(UINT8);
       _NUMPY_UNIFY_NOOP(UINT16);
       _NUMPY_UNIFY_NOOP(UINT32);
-      _NUMPY_UNIFY_PROMOTE(FLOAT32);
+      _NUMPY_UNIFY_PROMOTE_TO(FLOAT32, FLOAT64);
```

Comment:
Casting 32/64-bit integers to Float32 without a warning was pretty unsafe, that's what these changes are about. There are no tests to check, though, so we should probably open a JIRA to do that as risk mitigation